datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -994,8 +994,34 @@ pub async fn from_substrait_rel(
                     )
                     .await
                 }
-                _ => {
-                    not_impl_err!("Unsupported ReadType: {:?}", &read.as_ref().read_type)
+                Some(ReadType::ExtensionTable(ext)) => {
```

Comment:
I didn't actually get to handling ExtensionTables in https://github.com/apache/datafusion/pull/13803, but I think I've got most what we would need in place for it.

The way I would approach it, based on the work in that PR, is to add a method to the `SubstraitConsumer` trait like:
```rust
async fn consume_extension_table(&self, extension_table: &ExtensionTable) -> Result<LogicalPlan>
```
and wire it in here. Then, as a user you would be able to provide your own implementation of the decoder. This _might_ user the SerializerRegistry, but it doesn't necessarily need to.

Reply:
That's great! I like where that is going.
My goal here was to add the missing support for reading & writing extension tables leveraging only what's available (to keep the patch as small as possible).
But I do agree that uniform handling of Substrait extensions would make more sense and would overcome some of the current limitations.

I think the code here is really easy to migrate to the new `SubstraitConsumer` (and `SubstraitProducer`?) interfaces once they're available, by just replacing the `SerializerRegistry` calls with the new dedicated read/write methods. But the rest would be mostly unchanged by this migration.
FWIW, once your refactoring is complete, I think there would be no place for `SerializerRegistry` anymore, and it should be removed (or at least deprecated).

Reply:
> (and SubstraitProducer?)

I haven't gotten around to the producer yet, but if the SubstraitConsumer pattern makes sense to folks it should be easy enough to hammer it out.

> I think the code here is really easy to migrate to the new SubstraitConsumer

Would you be open to having https://github.com/apache/datafusion/pull/13803 merged first, and then porting your code over?

Reply:
Sure, I can do that.